tools/unicode_properties_parse/format_width_estimate_intervals.cpp

```diff
@@ -0,0 +1,209 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// The following code generates data for `_Width_estimate_intervals` in <format>.
+
+#include <charconv>
+#include <format>
+#include <fstream>
+#include <iostream>
+#include <regex>
+#include <vector>
+
+void _verify(bool test, int line, const char* msg) {
+    if (!test) {
+        std::cerr << "error at line " << line << ":" << msg << std::endl;
+        exit(EXIT_FAILURE);
+    }
+}
+#define verify(expr, msg) _verify((expr), __LINE__, (msg))
+static const char* impl_assertion_failed = "impl assertion failed";
+
+struct range_u {
+    uint32_t from, to;
+    range_u(uint32_t f, uint32_t t) : from(f), to(t) {}
+    range_u(uint32_t v) : from(v), to(v) {}
+};
+
+// The largest possible unicode won't exceed max_u.
+const uint32_t max_u = 0x1f'ffff;
+using table_u        = std::vector<bool>; // true: wide
+table_u make_table() {
+    return table_u(max_u + 1, false);
+}
+
+// For `_Width_estimate_intervals`.
+void print_intervals(const table_u& table) {
+    using namespace std;
+    cout << endl;
+    int c     = 0;
+    bool last = table[0];
+    for (uint32_t u = 0; u <= max_u; u++) {
+        if (table[u] != last) {
+            cout << "0x" << hex << uppercase << u << "u, ";
+            if (++c == 12) {
+                c = 0;
+                cout << endl;
+            }
+        }
+        last = table[u];
+    }
+    cout << endl;
+}
+
+void fill_range(table_u& table, const ::range_u rng, bool is_wide) {
+    const auto [from, to] = rng;
+    verify(from <= to && to <= max_u, impl_assertion_failed);
+    for (uint32_t u = from; u <= to; u++) {
+        table[u] = is_wide;
+    }
+}
+
+table_u get_table_cpp20() {
+    using namespace std;
+    const vector<range_u> std_wide_ranges_cpp20{
+        {0x1100, 0x115F},
+        {0x2329, 0x232A},
+        {0x2E80, 0x303E},
+        {0x3040, 0xA4CF},
+        {0xAC00, 0xD7A3},
+        {0xF900, 0xFAFF},
+        {0xFE10, 0xFE19},
+        {0xFE30, 0xFE6F},
+        {0xFF00, 0xFF60},
+        {0xFFE0, 0xFFE6},
+        {0x1F300, 0x1F64F},
+        {0x1F900, 0x1F9FF},
+        {0x20000, 0x2FFFD},
+        {0x30000, 0x3FFFD},
+    };
+    table_u table = make_table();
+    for (const range_u rng : std_wide_ranges_cpp20) {
+        fill_range(table, rng, true);
+    }
+    return table;
+}
+
+// Read data from a file with the same content as in https://www.unicode.org/Public/15.0.0/ucd/EastAsianWidth.txt

```

Comment:
If I'm correct, UAX #44 has evolving versions, with latest version being 15.0.0.

I'm confused by the decision between "latest" and "15.0.0". Which link should refer to here?
https://www.unicode.org/Public/15.0.0/ucd/EastAsianWidth.txt
vs
https://www.unicode.org/Public/UCD/latest/ucd/EastAsianWidth.txt

The annex in the paper(`"For the following code points, the estimated width used to be 1, and is 2 after the suggested
change"`) is also based on version 15.0.0. According to its wording, should we update test cases accordingly as new unicode versions come out?

Reply:
Existing py files refer to the latest version. So I think it'll be more consistent to use latest.

https://github.com/microsoft/STL/blob/8d18fec555022f4f355dabbb8e8c1893c91f5553/tools/unicode_properties_parse/grapheme_break_test_data_gen.py#L94-L96

https://github.com/microsoft/STL/blob/8d18fec555022f4f355dabbb8e8c1893c91f5553/tools/unicode_properties_parse/grapheme_break_property_data_gen.py#L283-L293

Reply:
> According to its wording, should we update test cases accordingly as new unicode versions come out?

Ideally, the new revision of C++ standard will refer to new version of Unicode. So I think we should update after the corresponding C++ paper is adopted into the working draft.

Reply:
This is required in C++23 see https://github.com/cplusplus/papers/issues/1429

Reply:
The current approach in the PR (link to latest, mention that this code was tested with v15) seems reasonable, resolving this comment.